rest/blip_api_test.go

```diff
@@ -0,0 +1,316 @@
+package rest
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/couchbase/go-blip"
+	"github.com/couchbaselabs/go.assert"
+)
+
+// This test performs the following steps against the Sync Gateway passive blip replicator:
+//
+// - Setup
+//   - Create an httptest server listening on a port that wraps the Sync Gateway Admin Handler
+//   - Make a BLIP/Websocket client connection to Sync Gateway
+// - Test
+//   - Verify Sync Gateway will accept the doc revision that is about to be sent
+//   - Send the doc revision in a rev request
+//   - Call changes endpoint and verify that it knows about the revision just sent
+//   - Call subChanges api and make sure we get expected changes back
+//
+// Replication Spec: https://github.com/couchbase/couchbase-lite-core/wiki/Replication-Protocol#proposechanges
+func TestBlipPushRevisionInspectChanges(t *testing.T) {
+
+	bt := CreateBlipTester(t, false)
+	defer bt.Close()
+
+	// Verify Sync Gateway will accept the doc revision that is about to be sent
+	var changeList [][]interface{}
+	changesRequest := blip.NewRequest()
+	changesRequest.SetProfile("changes")                             // TODO: make a constant for "changes" and use it everywhere
+	changesRequest.SetBody([]byte(`[["1", "foo", "1-abc", false]]`)) // [sequence, docID, revID]
+	sent := bt.sender.Send(changesRequest)
+	assert.True(t, sent)
+	changesResponse := changesRequest.Response()
+	assert.Equals(t, changesResponse.SerialNumber(), changesRequest.SerialNumber())
+	body, err := changesResponse.Body()
+	assertNoError(t, err, "Error reading changes response body")
+	err = json.Unmarshal(body, &changeList)
+	assertNoError(t, err, "Error unmarshalling response body")
+	assert.True(t, len(changeList) == 1) // Should be 1 row, corresponding to the single doc that was queried in changes
+	changeRow := changeList[0]
+	assert.True(t, len(changeRow) == 0) // Should be empty, meaning the server is saying it doesn't have the revision yet
+
+	// Send the doc revision in a rev request
+	revRequest := blip.NewRequest()
+	revRequest.SetCompressed(true)
+	revRequest.SetProfile("rev")
+	revRequest.Properties["id"] = "foo"
+	revRequest.Properties["rev"] = "1-abc"
+	revRequest.Properties["deleted"] = "false"
+	revRequest.Properties["sequence"] = "1"
+	revRequest.SetBody([]byte(`{"key": "val"}`))
```

Comment:
Once we start writing more tests, we'll be better off with helper methods to build each type of request.  That doesn't need to be in this initial PR, I suppose.

Reply:
Ok, yeah that crossed my mind, but I was debating it since I couldn't figure out how much value add the helper methods would add.  At some point I think the common patterns will emerge.